modules/openxr/extensions/openxr_debug_utils_extension.cpp

```diff
@@ -0,0 +1,281 @@
+		EXT_INIT_XR_FUNC(xrSetDebugUtilsObjectNameEXT);
+		EXT_INIT_XR_FUNC(xrSessionBeginDebugUtilsLabelRegionEXT);
+		EXT_INIT_XR_FUNC(xrSessionEndDebugUtilsLabelRegionEXT);
+		EXT_INIT_XR_FUNC(xrSessionInsertDebugUtilsLabelEXT);
+
+		debug_utils_ext = xrCreateDebugUtilsMessengerEXT_ptr && xrDestroyDebugUtilsMessengerEXT_ptr && xrSetDebugUtilsObjectNameEXT_ptr && xrSessionBeginDebugUtilsLabelRegionEXT_ptr && xrSessionEndDebugUtilsLabelRegionEXT_ptr && xrSessionInsertDebugUtilsLabelEXT_ptr;
+	}
+
+	// On successful init, setup our default messenger.
+	if (debug_utils_ext) {
+		int max_severity = GLOBAL_GET("xr/openxr/extensions/debug_utils");
+		XrDebugUtilsMessageSeverityFlagsEXT message_severities = 0;
+
+		if (max_severity >= 1) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_ERROR_BIT_EXT;
+		}
+		if (max_severity >= 2) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_WARNING_BIT_EXT;
+		}
+		if (max_severity >= 3) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT;
+		}
+		if (max_severity >= 4) {
+			message_severities |= XR_DEBUG_UTILS_MESSAGE_SEVERITY_VERBOSE_BIT_EXT;
+		}
+
+		XrDebugUtilsMessengerCreateInfoEXT callback_info = {
+			XR_TYPE_DEBUG_UTILS_MESSENGER_CREATE_INFO_EXT, // type
+			nullptr, // next
+			message_severities, // messageSeverities
+			XR_DEBUG_UTILS_MESSAGE_TYPE_GENERAL_BIT_EXT | XR_DEBUG_UTILS_MESSAGE_TYPE_VALIDATION_BIT_EXT | XR_DEBUG_UTILS_MESSAGE_TYPE_PERFORMANCE_BIT_EXT | XR_DEBUG_UTILS_MESSAGE_TYPE_CONFORMANCE_BIT_EXT, // messageTypes
+			&OpenXRDebugUtilsExtension::_debug_callback, // userCallback
+			nullptr, // userData
+		};
+
+		XrResult result = xrCreateDebugUtilsMessengerEXT(p_instance, &callback_info, &default_messenger);
+		if (XR_FAILED(result)) {
+			ERR_PRINT("OpenXR: Failed to create debug callback [" + OpenXRAPI::get_singleton()->get_error_string(result) + "]");
+		}
+
+		set_object_name(XR_OBJECT_TYPE_INSTANCE, uint64_t(p_instance), "Main Godot OpenXR Instance");
+	}
+}
+
+void OpenXRDebugUtilsExtension::on_instance_destroyed() {
+	if (default_messenger != XR_NULL_HANDLE) {
+		XrResult result = xrDestroyDebugUtilsMessengerEXT(default_messenger);
+		if (XR_FAILED(result)) {
+			ERR_PRINT("OpenXR: Failed to destroy debug callback [" + OpenXRAPI::get_singleton()->get_error_string(result) + "]");
+		}
+
+		default_messenger = XR_NULL_HANDLE;
+	}
+
+	xrCreateDebugUtilsMessengerEXT_ptr = nullptr;
+	xrDestroyDebugUtilsMessengerEXT_ptr = nullptr;
+	xrSetDebugUtilsObjectNameEXT_ptr = nullptr;
+	xrSessionBeginDebugUtilsLabelRegionEXT_ptr = nullptr;
+	xrSessionEndDebugUtilsLabelRegionEXT_ptr = nullptr;
+	xrSessionInsertDebugUtilsLabelEXT_ptr = nullptr;
+	debug_utils_ext = false;
+}
+
+void OpenXRDebugUtilsExtension::on_session_created(const XrSession p_instance) {
+	if (debug_utils_ext) {
+		set_object_name(XR_OBJECT_TYPE_SESSION, uint64_t(p_instance), "Main Godot OpenXR Session");
+
+		begin_debug_label_region("Godot session active");
```

Comment:
Does this extension need to be invoked first before all the others?

Reply:
After catching up with the code in `openxr_api`, it looks this should be invoked from `openxr_api` instead to ensure it runs before all the other extensions. Same thing for `set_object_name` in `on_instance_created`

Reply:
We can't, because at that point we haven't initialised it yet, it leads to a nice crash :P

We'd need some way to give priority to this extension over all the other extensions so it runs first.

Reply:
Owh this is the session call, I mixed up with instance because the parameter name is wrong, oops :) I'll fix that up.

I guess we could move it into openxr_api itself to ensure it runs ASAP.

Reply:
Ok changed this, but with the footnote that this requires a fix in the Meta XR Simulator before it works properly as the memory holding the string gets destroyed after the call ends.